airflow/contrib/hooks/gcp_vision_hook.py

```diff
@@ -0,0 +1,257 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import re
+
+from google.api_core.exceptions import GoogleAPICallError, AlreadyExists, RetryError
+from google.protobuf.json_format import MessageToJson
+
+from airflow import AirflowException
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+from google.cloud.vision_v1 import ProductSearchClient
+
+PARENT_TEMPLATE = 'projects/{}/locations/{}'
+PRODUCT_SET_NAME_TEMPLATE = 'projects/{}/locations/{}/productSets/{}'
+PRODUCT_NAME_TEMPLATE = 'projects/{}/locations/{}/products/{}'
+
+
+class CloudVisionHook(GoogleCloudBaseHook):
+    """
+    Hook for Google Cloud Vision APIs.
+    """
+
+    _client = None
+
+    def __init__(self,
+                 gcp_conn_id='google_cloud_default',
+                 delegate_to=None):
+        super(CloudVisionHook, self).__init__(gcp_conn_id, delegate_to)
+
+    def get_client(self):
+        if not self._client:
+            self._client = ProductSearchClient()
+        return self._client
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def create_product_set(self, location, product_set, project_id=None,
+                           product_set_id=None, retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetCreateOperator`
+        """
+        client = self.get_client()
+        parent = PARENT_TEMPLATE.format(project_id, location)
+        self.log.info('Creating a new ProductSet under the parent: %s', parent)
+        response = self._handle_request(lambda **kwargs: client.create_product_set(**kwargs), parent=parent,
+                                        product_set=product_set, product_set_id=product_set_id, retry=retry,
+                                        timeout=timeout, metadata=metadata)
+        self.log.info('ProductSet created:\n%s', response)
+
+        if not product_set_id:
+            # Product set id was generated by the API
+            name = ''
+            try:
+                name, product_set_id = self._get_autogenerated_id(response)
```

Comment:
This function is called twice. In both cases it is surrounded by a similar code. Can it be part of the function?